tools/mdbook_preprocessor/mdbook_preprocessor.js

```diff
@@ -0,0 +1,151 @@
+ * We use it to process code blocks with language "command-summary".
+ *
+ * @see processCommandSummary
+ *
+ * mdBook preprocessor documentation:
+ * https://rust-lang.github.io/mdBook/for_developers/preprocessors.html
+ */
+
+if (process.argv.length > 2) {
+  handleSupports();
+} else {
+  await handlePreprocess();
+}
+
+function handleSupports() {
+  if (process.argv[2] === "supports" && process.argv[3] === "html") {
+    process.exit(0);
+  }
+  process.exit(1);
+}
+
+/**
+ * An mdBook preprocessor receives JSON data `[context, book]` from stdin,
+ * modifies `book`, and writes `book` back to stdout.
+ *
+ * Example input:
+ * https://rust-lang.github.io/mdBook/for_developers/preprocessors.html#:~:text=mod%20test
+ */
+async function handlePreprocess() {
+  // Read from stdin
+  let stdin = "";
+  for await (const chunk of process.stdin) {
+    stdin += chunk;
+  }
+
+  // We don't care about the context. Only process the book.
+  const [, book] = JSON.parse(stdin);
+
+  processBook(book);
+
+  // Write to stdout
+  const output = JSON.stringify(book);
+  process.stdout.write(output + "\n");
+}
+
+/**
+ * @param {Book} book
+ */
+function processBook(book) {
+  for (const bookItem of book.sections) {
+    processBookItem(bookItem);
+  }
+}
+
+/**
+ * @param {BookItem} bookItem
+ */
+function processBookItem(bookItem) {
+  // bookItem is { "Chapter": Chapter }, "Separator", or { "PartTitle": string }
+  if (bookItem === "Separator") {
+    return;
+  }
+  // bookItem is { "Chapter": Chapter } or { "PartTitle": string }
+  if ("PartTitle" in bookItem) {
+    return;
+  }
+  // bookItem is { "Chapter": Chapter }
+
+  const chapter = bookItem.Chapter;
+  processChapter(chapter);
+}
+
+/**
+ * @param {Chapter} chapter
+ */
+function processChapter(chapter) {
+  for (const subItem of chapter.sub_items) {
+    processBookItem(subItem);
+  }
+
+  chapter.content = processContent(chapter.content);
+}
+
+/**
+ * @param {string} content
+ * @returns {string}
+ */
+function processContent(content) {
+  return content.replaceAll(/```command-summary\n([\s\S]*?)\n```/g, (_, code) => {
+    return processCommandSummary(code);
+  });
+}
+
+/**
+ * This function processes code blocks with language "command-summary".
+ *
+ * For example:
+ *
+ * ```command-summary
+ * git town append [--prototype] <branch-name>
+ * ```
+ *
+ * will become:
+ *
+ * <pre><code><div class="gt-command-summary" style="padding-left: 16ch; text-indent: -16ch"
+ *   ><span class="gt-command">git town append</span
+ *   > <span class="gt-argument">[--prototype]</span
+ *   > <span class="gt-argument">&lt;branch-name&gt;</span
+ *   ></div></code></pre>
+ *
+ * We can then apply custom styling in head.hbs.
+ *
+ * @param {string} code
+ * @returns {string}
+ */
+function processCommandSummary(code) {
+  return `<pre><code>${
+    code
+      .split("\n")
+      .map(line => {
+        const command = line.match(/^.*?(?=$| [\[<])/)?.[0];
```

Comment:
Would the regex still work if you omit the `^.*` at the beginning? It seems like the main goal here is to capture something in the middle of the line, so matching the start of the line and then allowing arbitrary characters afterwards might be unnecessary.

Reply:
No. The goal is to capture from the beginning of the line until the end or until the first parameter. It uses a [lookahead assertion](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Lookahead_assertion) (`(?=...)`) to match but not consume `$` (end of line) or `␣[\[<]`. So the part that gets returned is the text matched by `.*?`, with the lookahead assertion to force the `.*?` to consume as much as I want it to.

Reply:
I'll add a code comment explaining this.

Reply:
Thanks for the explanation, looks good to me now!